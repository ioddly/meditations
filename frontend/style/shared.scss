// TODO: Only use necessary CSS components
@import "primer-base/index.scss";
@import "primer-box/index.scss";
@import "primer-buttons/index.scss";
@import "primer-forms/index.scss";
@import "primer-navigation/index.scss";
@import "primer-layout/index.scss";
@import "primer-tooltips/index.scss";
@import "primer-support/index.scss";
@import "primer-utilities/index.scss";

@import "~medium-editor/dist/css/medium-editor.min.css";
@import "~medium-editor/dist/css/themes/bootstrap.min.css";
@import "~medium-editor-tables/dist/css/medium-editor-tables.css";

@import "~octicons/octicons/octicons.css";

@import "animations.scss";
@import "datepicker.scss";

$linkColor: #09d;
$bodyBg: white;
/* Slightly muted color of some UI actions e.g. task modifiers */
$uiColorMuted: #586069;
// Width at which to put the UI into mobile mode
$mobileWidth: 1024px;

.masthead {
  background-color: #f5f5f5;
}

.masthead .tabnav span {
  padding-right:5px;
}

.tabnav .selected {
  border-top: 2px solid #e36209;
  font-weight: bold;
}

#mobile-menu {
  display: none !important;
  position: fixed;
  right: 10px;
  top: 50px;
  z-index: 1;
}

@media (max-width: $mobileWidth) {
  #mobile-menu {
    display: flex !important;
  }
}

///// Notification bar
#notifications {
  animation: shake 0.82s cubic-bezier(.36,.07,.19,.97) both;
  transform: translate3d(0, 0, 0);
  backface-visibility: hidden;
  perspective: 1000px;

  position: fixed;
  left: 20px;
  top: 50px;
  z-index: 1;
  width: 320px;
}

///// MODAL PROMPT
#modal {
  position: fixed;
  z-index: 1;
  top: 20%;
  left: 50%;
  width: 350px;
  margin-left: -175px;
}

.no-display {
  display:none !important;
}

.btn-octicon {
  background: transparent;
  border:0;
  color: $uiColorMuted;
  padding: 0;
}

button.btn-octicon svg {
  fill: #09d;
}

.btn-danger svg {
  fill: #b33630;
}