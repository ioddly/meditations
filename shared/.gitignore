/node_modules
*.d.ts
*.js
*.d.ts.map
