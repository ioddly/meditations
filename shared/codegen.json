{
  "schema": "./schema.graphql",
  "documents": "./operations.graphql",
  "config": {
    "withHooks": true
  },
  "generates": {
    "../frontend/src/api/client.tsx": {
      "plugins": ["typescript", "typescript-operations", "typescript-urql"]
    },
    "./types-generated.ts": {
      "plugins": ["typescript"]
    }
  }
}
