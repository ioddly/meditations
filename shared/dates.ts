// dates.ts - shared date functionality

import { format, parse } from 'date-fns';

/**
 * Format a date to send over the wire, suitable for insertion into postgres
 */
export const formatWireDate = (date: Date) => {
  return format(date, "yyyy-MM-dd HH:mm:ssx");
}

const ref = new Date();

/**
 * Parse a date sent over the wire, suitable for insertion into postgres
 * @param {string} date 
 */
export const parseWireDate = (date: string) => {
  return parse(date, 'yyyy-MM-dd HH:mm:ssx', ref);
}
