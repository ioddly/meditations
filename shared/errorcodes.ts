// Stable error codes

/**
 * Error while parsing document
 */
export const SERVER_DOCUMENT_PARSE_ERROR = 'server/document-parse';

/**
 * Revision invariant broken (something went wrong with autosave)
 */
export const SERVER_REVISION_INVARIANT_BROKEN_ERROR = 'server/revision-invariant-broken';

/**
 * Database error
 */
export const SERVER_DATABASE_ERROR = 'server/database';

export type ErrorCode =
  | typeof SERVER_DOCUMENT_PARSE_ERROR
  | typeof SERVER_REVISION_INVARIANT_BROKEN_ERROR
  | typeof SERVER_DATABASE_ERROR
  ;