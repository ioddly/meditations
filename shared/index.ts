// types.ts - Document types

/**
 * For simplicty's sake, we keep these separate from SlateJS types, since SlateJS has some
 * fairly generous type definitions (e.g. allowing [key: string]: unknown) and we can narrow
 * our types down and use them in and around the program.
 * 
 * Also prefix them with T since there's a lot of overlap with both Slate types and browser builtins.
 */

export type TText = {
  text: string;
};

export type TElement = {
  type: 'line' | 'heading';
  children: TNode[];
};

export type TNode = TText | TElement;

export type NoteBody = TNode[];

/**
 * A document record; would be a toplevel database record containing an actual
 * document as well as information about it
 */
export type NoteRecord = {
  note_id: string;
  document: NoteBody;
};

export type TCollection = {
  collectionType: 'simple';
  name: string;
};

export type TCollectionEntry = {
  id: string;
  entryStatus: 'unset' | 'complete' | 'incomplete';
  entryValue: any;
};