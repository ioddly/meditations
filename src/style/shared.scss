// TODO: Only use necessary CSS components
@import "primer-css/index.scss";

/*
@import "~medium-editor/dist/css/medium-editor.min.css";
@import "~medium-editor/dist/css/themes/bootstrap.min.css";
@import "~medium-editor-tables/dist/css/medium-editor-tables.css";
@import "~react-datepicker/dist/react-datepicker";
@import "~react-tabs/style/react-tabs.css";

*/
/*
@import "~octicons/octicons/octicons.css";
*/

@import "animations.scss";

$linkColor: #09d;
$bodyBg: #fff;
$commentFont: $mono-font;

.masthead {
  background-color: #f5f5f5;
}


.masthead .tabnav span {
  padding-right:5px;
}

.tabnav .selected {
  border-top: 2px solid #e36209;
  font-weight: bold;
}

#mobile-menu {
  display: none !important;
  position: fixed;
  right: 10px;
  top: 50px;
  z-index: 1;
}

@media (max-width: 768px) {
  #mobile-menu {
    display: flex !important;
  }
}

///// Notification bar
#notifications {
  animation: shake 0.82s cubic-bezier(.36,.07,.19,.97) both;
  transform: translate3d(0, 0, 0);
  backface-visibility: hidden;
  perspective: 1000px;

  position: fixed;
  left: 20px;
  top: 50px;
  z-index: 1;
  width: 320px;
}
