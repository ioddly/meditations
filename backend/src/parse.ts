// parse.ts - document parsing

import { NoteBody, TagNote } from "../../shared"
import knex from "./knex"

/**
 * Return all relations that already exist in the database
 */
const getExistingRelations = (noteId: string) => {
  return knex.from('tag_notes').select<TagNote[]>('noteId, tagId').where({ 'noteId': noteId }).then(rows => rows);
}
// const discoverRelations = ()

type RelationTag = {
  type: 'tag',
  tagId: string,
  path: number[],
};

/**
 * Discover all relations that should exist as a result of a document
 */
export const discoverRelations = (noteBody: NoteBody, path: number[] = []): RelationTag[] => {
  let relations: RelationTag[] = [];

  for (let i = 0; i !== noteBody.length; i += 1) {
    const node = noteBody[i];

    switch (node.type) {
      case 'text': continue;
      case 'tag': {
        relations.push({ type: 'tag', tagId: node.tagId, path: [...path, i] });
      }
      case 'line': {
        relations = relations.concat(discoverRelations(node.children, [...path, i]));
      }
    }
  }

  return relations;
}