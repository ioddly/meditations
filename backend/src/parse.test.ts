import { NoteBody } from '../../shared';
import { discoverRelations } from './parse';

const sampleDoc1: NoteBody = [
  {
    type: 'line',
    children: [
      {
        type: 'text',
        text: 'This is a tagged document'
      },
      {
        type: 'tag',
        tagId: 'tag-3nntmoG2zhqHwt8EZamMiK',
        children: [
          {
            'type': 'text',
            'text': '',
          },
        ]
      }
    ],
  }
]

describe('discoverRelations', () => {
  it('successfully discovers basic relations', () => {
    const relations = discoverRelations(sampleDoc1);

    expect(relations).toMatchSnapshot();
  });
});