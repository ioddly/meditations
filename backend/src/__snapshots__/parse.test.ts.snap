// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`discoverRelations successfully discovers basic relations 1`] = `
Array [
  Object {
    "path": Array [
      0,
      1,
    ],
    "tagId": "tag-3nntmoG2zhqHwt8EZamMiK",
    "type": "tag",
  },
]
`;
